File: 992-subarrays-with-k-different-integers/992-subarrays-with-k-different-integers.py
class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        return self.withAtmostK(nums,k)-self.withAtmostK(nums,k-1)
    
    def withAtmostK(self,nums,k):
        i = 0; ans = 0;
        count = collections.Counter()
        for j in range(len(nums)):
            if(count[nums[j]]==0):
                k-=1
            count[nums[j]]+=1
            while(k<0):
                if(count[nums[i]]==1):
                    k+=1
                count[nums[i]]-=1
                i+=1
            ans+= j-i+1
        return ans

File: 992-subarrays-with-k-different-integers/992-subarrays-with-k-different-integers.cpp
class Solution {
public:
    
    int subarrWithatmostK(vector<int>& nums, int k, int n){
        unordered_map<int,int>umap;
        int i = 0, ans = 0;
        for(int j=0;j<n;j++){
            if(umap[nums[j]]==0){
                umap[nums[j]]++; k--;
            }else{
                umap[nums[j]]++;
            }
            while(k<0){
                if(umap[nums[i]]==1){
                    k++;
                }
                umap[nums[i]]--;
                i++;
            }
            ans+= j-i+1;
        }
        return ans;
    }
    
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        int n = nums.size();
        return subarrWithatmostK(nums,k,n)-subarrWithatmostK(nums,k-1,n);
    }
};